### DISCUSSION TITLE
Does zstd:chunked support prefetching?

### DISCUSSION THREAD

**Discussion by viluon** [containers/podman#18655] [Q&A] [answered]
Some of the zstd:chunked PRs mention eStargz support and that the format is generally based on stargz. An important feature of eStargz for my side-project is [prefetching](https://github.com/containerd/stargz-snapshotter/blob/7c805c28820d4056aa5ff219dd89493929e971c8/docs/estargz.md#prioritized-files-and-landmark-files), i.e., eagerly pulling certain files. Does zstd:chunked support prioritized files? Where can I learn more about the format?

**Comment by Luap99** [comment]
@giuseppe PTAL

**Comment by giuseppe** [comment, marked as answer]
@viluon no.  zstd:chunked doesn't allow yet to specify a priority for files.

What change would we need to perform to allow it?  Would it be enough to add a new field to the files index?

**Reply by viluon** [reply]
I believe so. eStargz prioritizes all files up to a "prefetch landmark" according to their order in the layer. The files index could have a field indicating the number of files to be prefetched. Conforming implementations of zstd:chunked could then order the files when building the layer (filling in the prefetch field accordingly) and pull the prefetch subset eagerly.

**Reply by giuseppe** [reply]
instead of a boolean, I'd prefer implementing it as two float values, that specifies what is the likelihood that a file is going to be needed at startup and at runtime by the container. so we can profile the container and see with different configurations when a file is needed before the container gets into a ready state, and when a file is needed before the container terminates.

With this flexibility, we can tweak the remote pull to sort the files, or define a threshold to decide whether a file must be preloaded or not.

**Reply by viluon** [reply]
I think that's a good idea, but I also think that eStargz supports prefetching files with a single range request (I *think* I heard that somewhere but I'm not sure). Of course, this isn't that much of a concern for the specification, but maybe the docs should point out that builders/pullers should optimise for this common case and order the files accordingly.

**Reply by polarathene** [reply]
SOCI snapshotter has a PR recently and associated issue discussion about implementing prefetch support. Although not entirely relevant to Podman I guess? (_I know of a third-party implementation for Podman that was based off the eStargz support, but that seems to be unmaintained now and stuck on SOCI v1_)

SOCI doesn't have `zstd:chunked` support implemented yet either, so just chiming in due to timing of their prefetch implementation if it's beneficial to try standardize more.
